pkg/tidb/forwarder.go

```diff
@@ -47,27 +60,71 @@ func NewForwarderConfig(cfg *config.Config) *ForwarderConfig {
 	return &ForwarderConfig{
 		TiDBRetrieveTimeout: time.Second,
 		TiDBTLSConfig:       cfg.TiDBTLSConfig,
+		CheckInterval:       cfg.CheckInterval,
 	}
 }
 
 type Forwarder struct {
 	ctx        context.Context
 	config     *ForwarderConfig
 	etcdClient *clientv3.Client
+	// The key is str label and value is the proxy
+	proxies      sync.Map
+	tidbPort     int
+	statusPort   int
+	donec        chan struct{}
+	pollInterval time.Duration
 }
 
-// TODO: Requires load balancing and health checks.
 func (f *Forwarder) Open() error {
-	// Currently this function does nothing.
+	var (
+		cluster []*tidbServerInfo
+		err     error
+	)
+	for {
+		cluster, err = f.getServerInfo()
+		if err == nil {
+			break
+		}
+		log.Error("Fail to get server info", zap.Error(err))
+		// TODO: config this or just use CheckInterval ?
+		<-time.After(time.Second * 5)
+	}
```

Comment:
In the case when TiDB / PD is not started, this function is trapped in the loop, so that it will block other constructors. I suspect this behaviour will cause many bugs (in other places)..

Reply:
What about using a goroutine to serve the codes above and add a flag to indicate whether the forwarder is ready? And if the forwarder is not ready, then any requests to the forwarder should be rejected.

Reply:
LGTM. Also what will happen if after some time all TiDBs are down? Will the `127.0.0.1:proxyPort` connection being blocked and waited, or being rejected instantly (the same behaviour as you described)?

Reply:
Under this circumstance, no `remote` will be picked on serving every connection and the request shall be rejected instantly.

Reply:
I found out that I can just use `pollingForTiDB` directly...